add class Main,Human,FamilyThree

SergeyCherevichenko opened the pull request:
Класс Human 
- создали поля класса
- создали два конструктора для создания детей(без списка) и родителей(со списком)
- добавили геттеры
- переопределили метод toString  класса Human
-
Класс FamilyThree
- создали поле List<Human>
- В конструкторе проинициализировали List как ArrayList
- Cоздали метод для добавления Human
- Переопределили метод toString класса FamilyThree
Создали тип enum gender, для ограничения ввода

Main 
-создали экземпляры классов и вывели FamilyThree

Liberate520 committed:
add class Main,Human,FamilyThree

Liberate520 committed:
add class Main,Human,FamilyThree

SergeyCherevichenko commented:
Спасибо

SergeyCherevichenko commented:
Спасибо.

SergeyCherevichenko committed:
add findChildren in FamilyThree

SergeyCherevichenko commented:
Добавил в FamilyThree метод findChildren(String lastname,String name). Изменения сохранились в 3780f54.  
видать что-то не то нажал когда последний commit делал

Liberate520 commented on src/FamilyTree.java:
```diff
@@ -0,0 +1,68 @@
+        return builder.toString();
+    }
+
+    public String findChildren(String name,String lastName) {
```
Как то содержание метода не соответствует названию. Вы ищите детей или человека?

Liberate520 commented on src/Human.java:
```diff
@@ -0,0 +1,72 @@
+    private String name;
+    private String lastName;
+    private Gender gender;
+    private LocalDate birthDay;
```
И дата смерти

Liberate520 commented on src/Human.java:
```diff
@@ -0,0 +1,72 @@
+
+    public List<Human> children;
+// Конструктор без детей
+    public Human(String name, String lastName, Gender gender, LocalDate birthDay) {
+        this.name = name;
+        this.lastName = lastName;
+        this.gender = gender;
+        this.birthDay = birthDay;
+
+    }
+    // Конструктор с детьми(списком)
+
+    public Human(String name, String lastName, Gender gender, LocalDate birthDay, List<Human> children) {
+        this.name = name;
+        this.lastName = lastName;
+        this.gender = gender;
+        this.birthDay = birthDay;
+        this.children = new ArrayList<>(children);
+
+    }
```
Нарушение принципа DRY

Liberate520 commented on src/Human.java:
```diff
@@ -0,0 +1,72 @@
+    public LocalDate getBirthDay() {
+        return birthDay;
+    }
+
```
Нет метода для добавления нового ребенка

SergeyCherevichenko committed:
1. добавлен дата смерти в поле HUman 2.Исправлен принцип работы метода findChild 3. Коеструкторы класса Human передалал по принципу DRY 4. В класс FamilyTree добавлен метод добавления ребенка  5. Доработан порядок вывода в консоль

SergeyCherevichenko commented:
1. добавлен дата смерти в поле HUman 2.Исправлен принцип работы метода findChild 3. Коеструкторы класса Human передалал по принципу DRY 4. В класс FamilyTree добавлен метод добавления ребенка  5. Доработан порядок вывода в консоль

SergeyCherevichenko commented:
Над заданием второго семинара сейчас работаю.
Спасибо

SergeyCherevichenko committed:
Подкоректировал метод addChild (как попросила IntelIdea)

SergeyCherevichenko commented:
Подкоректировал метод addChild (как попросила IntelIdea)

SergeyCherevichenko committed:
add interface Writable  add class FileHandler test metod save() and read()

SergeyCherevichenko commented:
Распределил поект по package.
Создал interface Writable с методами save() и read()
Cоздал класс FileHandler c с интерфейсом Writable  и реализовал метод save() и read()
Классы Human и FamilyTree наследуют интерфейс Serializable для сериализации и возможности записи и чтения

Протестировал запись и чтения - работает корректно. (Сохраняет все до того места где запучтили метод save(); Читает корректно. Пробовал как на desktop так и в папку IntelIdea.

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/FileHandler.java:
```diff
@@ -0,0 +1,30 @@
+            objectOutputStream.close();
+    }
+    // Метод для чтения
+    public void read() throws IOException, ClassNotFoundException {
```
Метод загрузки должен возвращать ранее сохраненный объект, чтобы мы имели возможность продолжить с ним работу. Вывод в консоль по большей части бесполезен)

SergeyCherevichenko committed:
Изменил метод read() на readeble () c возвпращаемым типом данных

SergeyCherevichenko commented:
Изменил в иньрфейсе Writable метод  void read() на Object readable()-возвращаемым типом данных.
Спасибо

SergeyCherevichenko committed:
add ServiceTree,add HumanBuilder,add HumanIterator,add 4 class comparator

SergeyCherevichenko commented:
Изменил структуру проекта с помощью класса высшего программаривония ServiceTree добавил Iterator для работы классе 
ServiceTree , добавил 4 метода сортировки.
Сейчас изменяю вид вывода на консоль, чтобы не выводились дети детей, думаю так будет симпатичнее
Спасибо.

SergeyCherevichenko committed:
add Writable(случайно удалил когда распределял по package

SergeyCherevichenko committed:
I made a more understandable output to the console  fixed the names of some sorting methods in ServiceTree (coincided with the names in FamilyTree

SergeyCherevichenko committed:
put the code in order

SergeyCherevichenko commented:
Сделал более приятный вывод в консоль. 
Изменил названия в методах сортировки в классе ServiceTree( были некоторые совпадения с названием методов в классе FamilyTree)
Навел порядок в коде

Спасибо.

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/familyTree/FamilyTree.java:
```diff
@@ -0,0 +1,51 @@
+        return new HumanIterator(humans);
+    }
+    public void sortByName(){
+        Collections.sort(humans,new HumanComparatorSortByName());
```
строчки можно свернуть в более короткий вариант

SergeyCherevichenko committed:
change project for generic

SergeyCherevichenko committed:
remove Dog Class(used for test)

SergeyCherevichenko commented:
Изменил проект под обобщения.
Спасибо.

SergeyCherevichenko committed:
change interface FamilyTreeItem

SergeyCherevichenko committed:
add toString for FamilyTree.class

SergeyCherevichenko commented:
вернул метод toString в FamilyTree class

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/file_handler/FileHandler.java:
```diff
@@ -0,0 +1,28 @@
+    }
+    // интерфейс, для того чтобы пользоваться классами записи и чтения
+    // методы для записи и чтения
+    public static interface Writable {
```
Интерфейсы также описываются в отдельном файле

Liberate520 replied:
А у вас и так описано в отдельном. Значите удалите отсюда интерфейс

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/familyTree/FamilyTree.java:
```diff
@@ -0,0 +1,112 @@
+    }
+
+    public void sortByName() {
+        Collections.sort(humans, new HumanComparatorSortByName());
```
при создании объектов использующих дженерики правильно добавлять \<>

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/familyTree/FamilyTree.java:
```diff
@@ -0,0 +1,112 @@
+        Collections.sort(humans, new HumanComparatorSortByCountChildren());
+    }
+
+    public String toString() {
```
Слишком длинный метод. Нужно разбить на более мелкие

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/familyTree/FamilyTree.java:
```diff
@@ -0,0 +1,112 @@
+import java.time.LocalDate;
+import java.time.Period;
+import java.util.*;
+public class FamilyTree<E extends FamilyTreeItem>  implements Serializable, Iterable<E> {
```
FamilyTreeItem\<E>

SergeyCherevichenko committed:
change project for MVP project
- add View
- add Presentor
- add Model(Service Tree)

SergeyCherevichenko commented:
Сделал поправки по замечаниям.
Переделал проект под MVP архитектуру.
Спасибо.

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/presenter/Presenter.java:
```diff
@@ -0,0 +1,79 @@
+    public void sortByName(){
+        service.sortTreeByName();
+       String answer = service.getHumanInfo();
+       view.printAnswer(answer);
```
Нарушение принципа DRY. У вас уже есть метод getHumanInfo

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/model_family_tree/human_builder/service_tree/ServiceTree.java:
```diff
@@ -0,0 +1,294 @@
+                    findChild = true;
+                }
+            }
+        }
```
Методы поиска должны быть описаны в классе дерева

SergeyCherevichenko committed:
change project for MVP project
- add View
- add Presentor
- add Model(Service Tree)
change project principles SOLID

SergeyCherevichenko commented:
Сделал поправки по замечаниям. Убрал лишние методы и перенес методы работы древа только в древо.
Посмотрел по принципам SOLID :
1 Принцип единственной ответсвенности.  + (Каждый класс выполняет свою единсттвенную задачу);
2 Принцип открытости и закрытости. +(Не делали сильную связь, а реализовывали через интерфейсы по необходимости).
3 Принцип подстановки Барбары Лисков. + (добавили generic и нигде в коде не проверем соответсвие типов, нас не интересует тип при переборе а метод).
4 Принцип разделения интерфейса. + (Практически везде в итерфейсах в данном проекте 1-2 метода, нет лишних реализаций этих методов интефейсов у классов)
5. Принцип инверсии зависимостей. + (запись и чтения реализовали через класс и интерфейс, соответсвеноо запись и чтение у нас не зависит конкретно от прописанного метода и можно изменять формат при необходимости)

Ну это я так по расуждал.
Спасибо.

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/model_family_tree/human_builder/service_tree/ServiceTree.java:
```diff
@@ -0,0 +1,81 @@
+public class ServiceTree  implements Serializable{
+    private FamilyTree<Human> tree;
+    private HumanBuilder builder;
+    FileHandler fileHandler;
```
нарушает пятый принцип солид. Связь должна быть через интерфейс, который мы также уже создали

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/model_family_tree/human_builder/service_tree/ServiceTree.java:
```diff
@@ -0,0 +1,81 @@
+         return tree.getHumanInfo();
+    }
+    public String removeHuman(String name, String lastName) {
+    return tree.removeHuman(name,lastName);
```
используйте автоформатирование текста для расставления отступов

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/view/ConsoleUi.java:
```diff
@@ -0,0 +1,117 @@
+        work = false;
+    }
+    public void addHuman(){
+        String name = dataEntry.name("Введите имя: ");
+        String lastName = dataEntry.name("Введите фамилию: ");
```
Надо было два отдельных метода просто сделать и не передавать никакой текст

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/view/ConsoleUi.java:
```diff
@@ -0,0 +1,117 @@
+        }
+    }
+    public void restoreTree() {
+        try {
```
обработку ошибок лучше делать на стороне модели, ведь это ошибки в работе именно модели

SergeyCherevichenko committed:
add change ServiceTree, ConsoleUi,FileHandler

SergeyCherevichenko commented:
Сделал поправки по последним замечаниям.

SergeyCherevichenko commented:
Спасибо

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/model_family_tree/human_builder/service_tree/ServiceTree.java:
```diff
@@ -0,0 +1,98 @@
+    public ServiceTree() {
+        tree = new FamilyTree<>();
+        builder = new HumanBuilder();
+        writable = new  FileHandler(this.tree);
```
Это нарушает пятый принцип солид. Два решения данной проблемы показывал на 6 семинаре во второй задаче)

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/presenter/Presenter.java:
```diff
@@ -0,0 +1,81 @@
+
+    public void sortByName() {
+        service.sortTreeByName();
+        String answer = service.getHumanInfo();
+        view.printAnswer(answer);
```
Нарушение принципа DRY. У вас есть метод getHumanInfo

Liberate520 commented on src/ru/home_work/cherevichenko_sergey/presenter/Presenter.java:
```diff
@@ -0,0 +1,81 @@
+    }
+
+    public void read()  {
+        service.readable();
```
Странное название для метода

SergeyCherevichenko committed:
add change ServiceTree, ConsoleUi,FileHandler

SergeyCherevichenko committed:
add changes Servicetree

SergeyCherevichenko commented:
Сделал изменения согласно замечаниям. В классе ServiceTree запись сделал через интерфейс, который передается через конструктор. В конструкторе по умолчанию сделал чтобы запись шла через FileHandler.
Переименовал метод readeble на read.
Заменил название метода getHumanInfo на getInfo(правда не совсем понял да у меня есть метод getHumanInfo, но я же вызываю метод с таким же названием у модели) или я не правильно понял?

SergeyCherevichenko committed:
Create README.md
